snapshotter/devmapper/metadata.go

```diff
@@ -0,0 +1,314 @@
+	db *bolt.DB
+}
+
+func NewPoolMetadata(dbfile string) (*PoolMetadata, error) {
+	db, err := bolt.Open(dbfile, 0600, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	metadata := &PoolMetadata{db: db}
+	if err := metadata.ensureDatabaseInitialized(); err != nil {
+		return nil, errors.Wrap(err, "failed to initialize database")
+	}
+
+	return metadata, nil
+}
+
+func (m *PoolMetadata) ensureDatabaseInitialized() error {
+	return m.db.Update(func(tx *bolt.Tx) error {
+		if _, err := tx.CreateBucketIfNotExists(devicesBucketName); err != nil {
+			return err
+		}
+
+		if _, err := tx.CreateBucketIfNotExists(deviceIDBucketName); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// AddDevice saves device info to database.
+// Callback should be used to acquire device ID or to rollback transaction in case of error.
+func (m *PoolMetadata) AddDevice(ctx context.Context, info *DeviceInfo, fn DeviceIDCallback) error {
+	return m.db.Update(func(tx *bolt.Tx) error {
+		devicesBucket := tx.Bucket(devicesBucketName)
+
+		// Make sure device name is unique
+		if err := getObject(devicesBucket, info.Name, nil); err == nil {
+			return ErrAlreadyExists
+		}
+
+		// Find next available device ID
+		deviceID, err := getNextDeviceID(tx)
+		if err != nil {
+			return err
+		}
+
+		if err := fn(deviceID); err != nil {
+			return err
+		}
+
+		info.DeviceID = deviceID
+
+		if err := putObject(devicesBucket, info.Name, info, false); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func getNextDeviceID(tx *bolt.Tx) (int, error) {
+	bucket := tx.Bucket(deviceIDBucketName)
+	cursor := bucket.Cursor()
+
+	// Check if any device id can be reused.
+	// Bolt stores its keys in byte-sorted order within a bucket.
+	// This makes sequential iteration extremely fast.
+	for key, taken := cursor.First(); key != nil; key, taken = cursor.Next() {
+		isFree := taken[0] == deviceFree
+		if !isFree {
+			continue
+		}
+
+		id, err := strconv.Atoi(string(key))
+		if err != nil {
+			return 0, err
+		}
+
+		if err := markDeviceID(tx, id, deviceTaken); err != nil {
+			return 0, err
+		}
+
+		return id, nil
+	}
+
+	// Try allocate new device ID
+	seq, err := bucket.NextSequence()
+	if err != nil {
+		return 0, err
+	}
+
+	if seq >= maxDeviceID {
+		return 0, errors.Errorf("couldn't find free device key")
+	}
+
+	id := int(seq)
+	if err := markDeviceID(tx, id, deviceTaken); err != nil {
+		return 0, err
+	}
+
+	return id, nil
+}
+
+func markDeviceID(tx *bolt.Tx, deviceID int, state byte) error {
+	var (
+		bucket = tx.Bucket(deviceIDBucketName)
+		key    = strconv.Itoa(deviceID)
+		value  = []byte{state}
+	)
+
+	if err := bucket.Put([]byte(key), value); err != nil {
+		return errors.Wrapf(err, "failed to free device id %q", key)
+	}
+
+	return nil
+}
+
+// UpdateDevice updates device info in metadata store.
+// Callback should be used to modify device properties or to rollback update transaction.
+// Name and Device ID couldn't be changed and will be ignored.
```

Comment:
I'd like to see an explicit error if this condition is violated instead of ignoring.